packages/yew-router/src/utils.rs

```diff
@@ -0,0 +1,62 @@
+use std::collections::HashMap;
+use wasm_bindgen::{JsCast, JsValue};
+
+pub fn base_url() -> Option<String> {
+    match yew::utils::document().query_selector("base") {
```

Comment:
We can't use that because we need the base URI specified in `<base>` so we can build the URL like `/path/to/something` to be passed to history API.

On `https://examples.yew.rs/router/`
```js
document.querySelector('base').baseURI
// "https://examples.yew.rs/router/"
```

We can't use it as it also includes the hostname.

Reply:
Right, but removing the host from the URL shouldn't be a problem :P.
The advantage of reading the baseURI (which should be read from `document.baseURI` directly btw, not the `<base>` element) is that it already handles a lot of weird user input for us.
For instance, `base:href` allows relative paths, so something like `href="../../root"` is perfectly valid (although certainly not useful in any way) and `document.baseURI` will give us a "clean" path.

Reply:
Well, `document.baseURI` on this page returns `"https://github.com/yewstack/yew/pull/1791"`. For me, it has to be `"/"` (or with hostname, `"https://github.com/"`. We already have the entire path to be pushed onto the router, we just need base to handle cases like yew examples. It seems to me that using `href` from `<base>` is easier/cleaner than trying to modify `document.baseURI`. If you have a better solution then please let me know.

Reply:
You're right, I forgot that `document.baseURI` defaults to the entire URL if no `<base>` tag is present.
At this point, checking whether a `<base>` element is present first and then extracting the path from `document.baseURI` is even more convoluted than handling all the quirks of reading the `href` attribute manually.
So you're right, let's do it that way! But we do still need to make a few changes to make this correct:

- Only consider base elements with the `href` attribute (i.e. use this query selector `"base[href]"`)
- Use the [`href` getter](https://rustwasm.github.io/wasm-bindgen/api/web_sys/struct.HtmlBaseElement.html#method.href) instead of reading the raw attribute value. This gives us the normalized URL so we don't have to worry about relative paths and other things. It's an absolute URL though so we're gonna have to use `(new URL("<url>")).pathname` to get the pathname from it.

Some unresolved questions:
- how do we deal with absolute URLs with a host that doesn't match the current location? I think we shouldn't check for this case and use the path regardless
- what do we do about paths containing query parameters or anchors? Again, I think we should just naively ignore those.

Reply:
I agree with what you said about unsolved questions. We shouldn't check for these cases.

I went and make this change. I think we'll run into edge cases with respect to `base` when we improve the ergonomics for nested routers.